node/actors/bft/src/chonky_bft/timeout.rs

```diff
@@ -0,0 +1,197 @@
+        // malicious replica starts spamming messages for future views
+        let active_views: HashSet<_> = self.timeout_views_cache.values().collect();
+        self.timeout_qcs_cache
+            .retain(|view_number, _| active_views.contains(view_number));
+
+        // Now we check if we have enough weight to continue. If not, we wait for more messages.
+        if weight < self.config.genesis().validators.quorum_threshold() {
+            return Ok(());
+        };
+
+        // ----------- We have a QC. Now we process it. --------------
+
+        // Consume the created timeout QC for this view.
+        let timeout_qc = self.timeout_qcs_cache.remove(&message.view.number).unwrap();
+
+        // We update our state with the new timeout QC.
+        if let Some(commit_qc) = timeout_qc.high_qc() {
+            self.process_commit_qc(ctx, commit_qc)
+                .await
+                .wrap("process_commit_qc()")?;
+        }
+        self.high_timeout_qc = max(Some(timeout_qc.clone()), self.high_timeout_qc.clone());
+
+        // Metrics.
+        let now = ctx.now();
+        metrics::METRICS
+            .leader_commit_phase_latency
+            .observe_latency(now - self.phase_start);
+        self.phase_start = now;
+
+        // Start a new view.
+        self.start_new_view(ctx, message.view.number.next()).await?;
+
+        Ok(())
+    }
+
+    /// This blocking method is used whenever we timeout in a view.
+    pub(crate) async fn start_timeout(&mut self, ctx: &ctx::Ctx) -> ctx::Result<()> {
+        // Update the state machine.
+        self.phase = validator::Phase::Timeout;
+
+        // Backup our state.
+        self.backup_state(ctx).await.wrap("backup_state()")?;
+
+        // Broadcast our timeout message.
+        let output_message = ConsensusInputMessage {
+            message: self
+                .config
+                .secret_key
+                .sign_msg(validator::ConsensusMsg::ReplicaTimeout(
+                    validator::ReplicaTimeout {
+                        view: validator::View {
+                            genesis: self.config.genesis().hash(),
+                            number: self.view_number,
+                        },
+                        high_vote: self.high_vote.clone(),
+                        high_qc: self.high_commit_qc.clone(),
+                    },
+                )),
+        };
+
+        self.outbound_pipe.send(output_message.into());
+
+        // Log the event.
+        tracing::info!("Timed out at view {}", self.view_number);
+        metrics::METRICS.replica_view_number.set(self.view_number.0);
+
+        // Reset the timeout. This makes us keep sending timeout messages until the consensus progresses.
+        // However, this isn't strictly necessary since the network retries messages until they are delivered.
+        // This is just an extra safety measure.
+        self.timeout_deadline = time::Deadline::Finite(ctx.now() + Self::TIMEOUT_DURATION);
```

Comment:
Can you make it actually send the timeout once? This is not the responsibility of the chonky module to retry sending messages.

Reply:
I can. Note that we will need to keep the three phases, instead of reducing to two like we are discussing in Slack.

Reply:
I would vote for having 3 phases then.